VnodeCacheDesign.md

```diff
@@ -2,37 +2,83 @@
 
 ## Goals
 
-- Reduce the time required to determine if a file is inside of a virtualization root
+- Reduce the time required to determine if a file is inside a virtualization root
 - Reduce the time required to find a file's virtualization root
-- **TBD:** Eliminate the use of file flags to determine if a file is inside of a root
+- Eliminate the use of file flags to determine if a file is inside of a root
 
 ## Design 
 
 ### Cached Data
 
-PrjFSKext will cache the following information for each vnode:
+PrjFSKext will cache the following information for each `vnode_t` pointer:
 
 - The `vnode_t` itself
-- The vnode's `uint32_t` vid (generation number)
+- The vnode's `uint32_t` vid (vnode generation number)
+- A `uint32_t` virtualization\cache generation number (hereby referred to as `vrgid`).  If a cache entry's `vrgid` does not match its roots it means that all entries for that root have been invalidated and the `VirtualizationRootHandle` in the cache entry is no longer valid.  There will also be a global `vrgid`  for all vnodes that are not inside of any virtualization roots.
 - `VirtualizationRootHandle` for the vnode (or `RootHandle_None` if the vnode is not in a root).  **NOTE** This assumes that `VirtualizationRootHandle`  are not recycled as VFS4G instances are unmounted\remounted.
 
-The data for each vnode will be stored in a hash table that assumes there will be `desiredvnodes` total vnodes.
+### Hash Table
+
+The data for each vnode will be stored in a fixed-size hash table with the `vnode_t` pointer serving as the key.
+
+#### Collision Handling
+
+Index collisions will be handled using open addressing.  The exact technique (e.g. linear probing, double hashing, etc.) will be determined once we have sample sets of `vnode_t` pointers and can evaluate the quality of our hash function.
+
+#### Size
+
+The hash table will be a fixed size of length `desiredvnodes`.
```

Comment:
`desiredvnodes` is a target value; if I understand the code correctly, there can actually be a few more than that, as the recycling vnodes aren't counted towards the total. So it's not quite a fixed size, but "not much larger than `desiredvnodes`".

Reply:
> there can actually be a few more than that, as the recycling vnodes aren't counted towards the total

Good point.  The "fixed size" comment was because we're looking to avoid the cost of growing the cache if it there's not enough space.  It sounds like we might instead want to pick a size some percentage larger than `desiredvnodes` (e.g. `desiredvnodes * 1.25`).

Reply:
Yeah, we'll need to try and see how it behaves in practice; I think we should at least support resizing in case it becomes necessary (e.g. as a result of some other kext changing `desiredvnodes`), even if it introduces a brief performance stumble.

Reply:
@pmj good point, I'll update the doc